### DISCUSSION TITLE
Problems with c++ code we try to skip

### DISCUSSION THREAD

**Discussion by leap-pea** [golangci/golangci-lint#5117] [Q&A] [answered]
Hey,

I have troubles getting `golangci-lint` to work for our projects.

1) When I run it locally, I get the following error:

```
pea@peabox:bitmask-vpn golangci-lint run          
gui/backend.go:1: : # 0xacab.org/leap/bitmask-vpn/gui
handlers.cpp:1:10: fatal error: QTimer: No such file or directory
    1 | #include <QTimer>
      |          ^~~~~~~~
compilation terminated. (typecheck)
package main
```

In the config, I try to exclude this file:

```yaml
issues:
   exclude-dirs:
     - "gui"
```

If I understand the [FAQ](https://golangci-lint.run/welcome/faq/#why-is-it-not-possible-to-skipignore-typecheck-errors) correctly, the `typecheck` can not be skipped. So excluding does not work here, right? I also don't know how to fix this issue and what's missing here (I'm on the go part in this project).

2) If I run it in the [CI](https://0xacab.org/leap/bitmask-vpn/-/jobs/666359), I get

```
$ golangci-lint run
gui/backend.go:1: : exec: "g++": executable file not found in $PATH (typecheck)
package main
Cleaning up project directory and file based variables
01:23
ERROR: Job failed: exit code 1
```

We are using the `golangci/golangci-lint:v1.61-alpine` docker image. What is the intended way to fix that? Should we install `g++` in the container? We are still fine excluding this file from checking.

Help is really appreciated! Thanks

**Comment by ldez** [comment]
Hello,

> So excluding does not work here, right?

Excluding has never worked: most linters were not working and all the reports were ignored.

> What is the intended way to fix that? Should we install g++ in the container? 

As explained inside the [doc](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors), if using CGO, ensure all required system libraries are installed.

So yes you should install g++, or create a custom Docker image with `golangci/golangci-lint:v1.61-alpine` as the base image with g++.

Also locally (and inside the container) you have missing system libraries, so you must install them too.

In summary, you should install the same things that are required to build your program.

**Reply by leap-pea** [reply]
Thanks for the information.
I build the project regularly, so I'm wondering why it can't find the dependencies/headers. Any idea where to look for/debug?

**Reply by ldez** [reply]
I'm not a cgo expert but:

```console
$ docker run -it --rm -v $(pwd):/app -w /app golangci/golangci-lint:v1.61.0 sh
# go build -o foo ./gui
# 0xacab.org/leap/bitmask-vpn/gui
handlers.cpp:1:10: fatal error: QTimer: No such file or directory
    1 | #include <QTimer>
      |          ^~~~~~~~
compilation terminated.
# apt update
...
# apt-get install -y pkg-config g++ libqt6svg6-dev qt6-tools-dev qt6-tools-dev-tools qt6-base-dev libqt6qml6 qt6-declarative-dev dh-golang libgl-dev qt6-declarative-dev-tools qt6-l10n-tools
...
# apt-get install -y libqt6core6 libqt6gui6 libqt6qml6 libqt6widgets6 libstdc++6 libqt6svg6 qml6-module-qtquick qml6-module-qtquick-controls qml6-module-qtquick-dialogs qml6-module-qtquick-layouts qml6-module-qtqml-workerscript qml6-module-qtquick-templates qml6-module-qt-labs-settings qml6-module-qtquick-window qml6-module-qt-labs-platform qml6-module-qtcore qml6-module-qt5compat-graphicaleffects openvpn policykit-1-gnome python3 iptables
...
# go build -o foo ./gui
# 0xacab.org/leap/bitmask-vpn/gui
handlers.cpp:1:10: fatal error: QTimer: No such file or directory
    1 | #include <QTimer>
      |          ^~~~~~~~
compilation terminated.
# 
```

I think there are some missing flags or something else, I don't know.

**Reply by leap-pea** [reply]
When using `RELEASE=yes PROVIDER=bitmask make -j 4` in combination with `set -x`, I get:
> go build -buildmode=c-archive '-ldflags=-X 0xacab.org/leap/bitmask-vpn/pkg/config/version.appVersion=0.24.10-rc.1-1-gb62f9ca1 -extar=ar -extld=ld -extldflags=' -o lib/libgoshim.a gui/backend.go

Is there a way to tell that `golangci-lint`?

**Reply by ldez** [reply]
The minimal command is `go build -o foo gui/backend.go`.

so:
```
golangci-lint run ./gui/backend.go
```


<details>

```console
$ docker run -it --rm golangci/golangci-lint:v1.61.0 sh
# git clone -q https://0xacab.org/leap/bitmask-vpn.git && cd bitmask-vpn
# golangci-lint run -v ./gui/backend.go
INFO golangci-lint has version 1.61.0 built with go1.23.1 from a1d6c560 on 2024-09-09T17:44:42Z 
INFO [config_reader] Config search paths: [./ /go/bitmask-vpn/gui /go/bitmask-vpn /go / /root] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 6 linters: [errcheck gofmt gosimple govet ineffassign unused] 
INFO [loader] Go packages loading at mode 575 (files|imports|types_sizes|compiled_files|deps|exports_file|name) took 7.473415628s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 395.45µs 
INFO [linters_context/goanalysis] analyzers took 20.493102569s with top 10 stages: buildir: 18.329173032s, inspect: 914.456823ms, printf: 488.758076ms, ctrlflow: 470.139744ms, fact_purity: 285.673215ms, gofmt: 1.27035ms, unused: 755.018µs, S1038: 598.33µs, cgocall: 298.76µs, S1039: 259.142µs 
INFO [runner] Issues before processing: 49, after processing: 0 
INFO [runner] Processors filtering stat (in/out): cgo: 49/0 
INFO [runner] processing took 6.257µs with stages: cgo: 1.822µs, nolint: 581ns, invalid_issue: 391ns, max_same_issues: 376ns, skip_files: 353ns, skip_dirs: 284ns, filename_unadjuster: 263ns, uniq_by_line: 247ns, fixer: 246ns, max_from_linter: 174ns, exclude: 174ns, exclude-rules: 169ns, sort_results: 166ns, diff: 161ns, path_prettifier: 160ns, identifier_marker: 142ns, severity-rules: 138ns, autogenerated_exclude: 134ns, max_per_file_from_linter: 73ns, path_shortener: 68ns, source_code: 68ns, path_prefixer: 67ns 
INFO [runner] linters took 7.827794309s with stages: goanalysis_metalinter: 7.827754476s 
INFO File cache stats: 0 entries of total size 0B 
INFO Memory: 155 samples, avg is 439.5MB, max is 964.3MB 
INFO Execution took 15.303780151s                 
# 
```

</details>


Your application has several entry points:

<details>

```console
$ golangci-lint run ./pkg/... ./cmd/... ./branding/... ./tests/...               
cmd/snowflake-client/main.go:175:16: Error return value of `pt.ProxyError` is not checked (errcheck)
                pt.ProxyError("proxy is not supported")
                             ^
cmd/snowflake-client/main.go:187:20: Error return value of `pt.CmethodError` is not checked (errcheck)
                                pt.CmethodError(methodName, err.Error())
                                               ^
cmd/snowflake-client/main.go:195:19: Error return value of `pt.CmethodError` is not checked (errcheck)
                        pt.CmethodError(methodName, "no such method")
                                       ^
cmd/snowflake-client/main.go:71:4: S1023: redundant `return` statement (gosimple)
                        return
                        ^
cmd/snowflake-client/main.go:62:5: S1023: redundant `return` statement (gosimple)
                                return
                                ^
...
```

</details>

**Reply by ldez** [reply, marked as answer]
I found another trick, you can add the following build constraints inside the `.cpp` files inside `./gui`:

```go
//go:build cpp
```

If you put this comment on top of the following files every works as expected:

```
./gui
├── handlers.cpp
├── main.cpp
└── qjsonmodel.cpp
```

<details>

```console
$ golangci-lint run 
pkg/helper/ports.go:11:6: func `getFirstAvailablePortFrom` is unused (unused)
func getFirstAvailablePortFrom(port int) int {
     ^
pkg/helper/ports.go:23:6: func `isPortAvailable` is unused (unused)
func isPortAvailable(port int) bool {
     ^
pkg/helper/ports.go:33:6: func `writePortToFile` is unused (unused)
func writePortToFile(port int) error {
     ^
pkg/helper/ports.go:43:6: func `getExecutableDir` is unused (unused)
func getExecutableDir() (string, error) {
     ^
pkg/bitmask/autostart/autostart.go:33:2: const `errorMsg` is unused (unused)
        errorMsg = `An error has ocurred initializing the VPN: %v`
        ^
pkg/backend/api.go:75:22: Error return value of `ctx.bm.Reconnect` is not checked (errcheck)
                go ctx.bm.Reconnect()
                                   ^
pkg/backend/api.go:87:19: Error return value of `ctx.bm.Reconnect` is not checked (errcheck)
                ctx.bm.Reconnect()
                                ^
pkg/backend/api.go:108:22: Error return value of `ctx.cfg.SetUseObfs4` is not checked (errcheck)
                ctx.cfg.SetUseObfs4(true)
                                   ^
pkg/backend/api.go:109:20: Error return value of `ctx.cfg.SetUseKCP` is not checked (errcheck)
                ctx.cfg.SetUseKCP(false)
                                 ^
pkg/backend/api.go:111:22: Error return value of `ctx.cfg.SetUseObfs4` is not checked (errcheck)
                ctx.cfg.SetUseObfs4(true)
                                   ^
pkg/backend/api.go:112:20: Error return value of `ctx.cfg.SetUseKCP` is not checked (errcheck)
                ctx.cfg.SetUseKCP(true)
                                 ^
pkg/backend/api.go:114:22: Error return value of `ctx.cfg.SetUseObfs4` is not checked (errcheck)
                ctx.cfg.SetUseObfs4(false)
                                   ^
pkg/backend/api.go:115:20: Error return value of `ctx.cfg.SetUseKCP` is not checked (errcheck)
                ctx.cfg.SetUseKCP(false)
                                 ^
pkg/backend/api.go:124:19: Error return value of `ctx.cfg.SetUseUDP` is not checked (errcheck)
        ctx.cfg.SetUseUDP(udp)
                         ^
pkg/backend/api.go:133:25: Error return value of `ctx.cfg.SetUseSnowflake` is not checked (errcheck)
        ctx.cfg.SetUseSnowflake(snowflake)
                               ^
pkg/backend/api.go:134:21: Error return value of `ctx.bm.UseSnowflake` is not checked (errcheck)
        ctx.bm.UseSnowflake(snowflake)
                           ^
pkg/backend/webapi.go:22:11: Error return value of `w.Write` is not checked (errcheck)
                        w.Write([]byte("401 - Unauthorized"))
                               ^
pkg/backend/webapi.go:93:26: Error return value of `ctx.bm.SetTransport` is not checked (errcheck)
                        go ctx.bm.SetTransport(string(t))
                                              ^
pkg/backend/actions.go:28:6: func `getGateway` is unused (unused)
func getGateway() string {
     ^
pkg/backend/status.go:112:3: S1000: should use for range instead of for { select {} } (gosimple)
                for {
                ^
pkg/backend/status.go:121:2: S1000: should use for range instead of for { select {} } (gosimple)
        for {
        ^
pkg/backend/api.go:215:2: S1021: should merge variable declaration with assignment on next line (gosimple)
        var providerOpts *bitmask.ProviderOpts
        ^
pkg/backend/webapi.go:38:17: printf: non-constant format string in call to fmt.Fprintf (govet)
        fmt.Fprintf(w, ctx.Status.String())
                       ^
pkg/backend/webapi.go:42:17: printf: non-constant format string in call to fmt.Fprintf (govet)
        fmt.Fprintf(w, ctx.bm.GetCurrentGateway())
                       ^
pkg/backend/webapi.go:71:17: printf: non-constant format string in call to fmt.Fprintf (govet)
        fmt.Fprintf(w, string(locationJson))
                       ^
pkg/pickle/helpers.go:76:13: Error return value of `fmt.Scanln` is not checked (errcheck)
                fmt.Scanln(&confirm)
                          ^
cmd/snowflake-client/main.go:175:16: Error return value of `pt.ProxyError` is not checked (errcheck)
                pt.ProxyError("proxy is not supported")
                             ^
cmd/snowflake-client/main.go:187:20: Error return value of `pt.CmethodError` is not checked (errcheck)
                                pt.CmethodError(methodName, err.Error())
                                               ^
cmd/snowflake-client/main.go:195:19: Error return value of `pt.CmethodError` is not checked (errcheck)
                        pt.CmethodError(methodName, "no such method")
                                       ^
cmd/snowflake-client/main.go:71:4: S1023: redundant `return` statement (gosimple)
                        return
                        ^
cmd/snowflake-client/main.go:62:5: S1023: redundant `return` statement (gosimple)
                                return
                                ^
pkg/vpn/management/client.go:302:22: func `(*MgmtClient).sendCommandPayload` is unused (unused)
func (c *MgmtClient) sendCommandPayload(payload []byte) error {
                     ^
pkg/vpn/management/event.go:34:5: var `clientEventKW` is unused (unused)
var clientEventKW = []byte("CLIENT")
    ^
pkg/vpn/management/event.go:36:5: var `fatalEventKW` is unused (unused)
var fatalEventKW = []byte("FATAL")
    ^
pkg/vpn/management/event.go:38:5: var `infoEventKW` is unused (unused)
var infoEventKW = []byte("INFO")
    ^
pkg/vpn/management/event.go:39:5: var `logEventKW` is unused (unused)
var logEventKW = []byte("LOG")
    ^
pkg/vpn/management/event.go:40:5: var `needOkEventKW` is unused (unused)
var needOkEventKW = []byte("NEED-OK")
    ^
pkg/vpn/management/event.go:41:5: var `needStrEventKW` is unused (unused)
var needStrEventKW = []byte("NEED-STR")
    ^
pkg/vpn/management/event.go:42:5: var `passwordEventKW` is unused (unused)
var passwordEventKW = []byte("PASSWORD")
    ^
branding/motd-cli/main.go:51:15: printf: non-constant format string in call to fmt.Printf (govet)
                        fmt.Printf(t.Str)
                                   ^
pkg/vpn/openvpn.go:423:23: Error return value of `b.obfsvpnProxy.Stop` is not checked (errcheck)
                        b.obfsvpnProxy.Stop()
                                           ^
pkg/vpn/status.go:70:34: Error return value of `b.managementClient.SendPassword` is not checked (errcheck)
                b.managementClient.SendPassword(b.launch.MngPass)
                                               ^
pkg/vpn/status.go:71:36: Error return value of `b.managementClient.SetStateEvents` is not checked (errcheck)
                b.managementClient.SetStateEvents(true)
                                                 ^
pkg/vpn/utils.go:101:11: Error return value of `rand.Read` is not checked (errcheck)
        rand.Read(buff)
                 ^
pkg/vpn/bonafide/bonafide.go:269:30: Error return value of `b.fetchGatewaysFromMenshen` is not checked (errcheck)
                        b.fetchGatewaysFromMenshen()
                                                  ^
pkg/vpn/bonafide/bonafide.go:312:26: Error return value of `b.gateways.setUserChoice` is not checked (errcheck)
        b.gateways.setUserChoice(label)
                                ^
pkg/vpn/bonafide/eip_service.go:124:17: Error return value of `resp.Body.Read` is not checked (errcheck)
                resp.Body.Read(buf)
                              ^
pkg/vpn/bonafide/eip_service.go:272:16: Error return value of `json.Unmarshal` is not checked (errcheck)
        json.Unmarshal([]byte(byteValue), &cfg)
                      ^
pkg/vpn/bonafide/bonafide.go:60:2: field `snowflake` is unused (unused)
        snowflake         bool
        ^
pkg/vpn/bonafide/eip_service.go:20:2: field `defaultGateway` is unused (unused)
        defaultGateway       string
        ^
pkg/vpn/bonafide/eip_service.go:28:2: field `defaultGateway` is unused (unused)
        defaultGateway       string
        ^
pkg/vpn/bonafide/gateways.go:204:23: func `(*gatewayPool).getGatewayByHost` is unused (unused)
func (p *gatewayPool) getGatewayByHost(host string) (Gateway, error) {
                      ^
pkg/vpn/bonafide/bonafide_test.go:44:6: type `nopCloser` is unused (unused)
type nopCloser struct {
     ^
pkg/vpn/bonafide/bonafide_test.go:48:18: func `nopCloser.Close` is unused (unused)
func (nopCloser) Close() error { return nil }
                 ^
pkg/vpn/bonafide/bonafide.go:220:3: S1000: should use for range instead of for { select {} } (gosimple)
                for {
                ^
pkg/vpn/bonafide/bonafide.go:138:2: S1008: should use 'return err != nil' instead of 'if err != nil { return true }; return false' (gosimple)
        if err != nil {
        ^
pkg/vpn/bonafide/bonafide.go:368:12: ineffectual assignment to err (ineffassign)
        dataJSON, err := ioutil.ReadAll(resp.Body)
                  ^
pkg/vpn/bonafide/gateways.go:351:3: ineffectual assignment to distance (ineffassign)
                distance := 13
                ^
pkg/vpn/bonafide/gateways_test.go:70:7: ineffectual assignment to err (ineffassign)
        gws, err := pool.getRandomGatewaysByLocation("a", "openvpn")
             ^
pkg/snowflake/lib/lib_test.go:113:13: Error return value of `p.Collect` is not checked (errcheck)
                        p.Collect()
                                 ^
pkg/snowflake/lib/lib_test.go:114:13: Error return value of `p.Collect` is not checked (errcheck)
                        p.Collect()
                                 ^
pkg/snowflake/lib/lib_test.go:115:13: Error return value of `p.Collect` is not checked (errcheck)
                        p.Collect()
                                 ^
pkg/snowflake/lib/lib_test.go:167:11: Error return value is not checked (errcheck)
                        Handler(socks, d)
                               ^
pkg/snowflake/lib/webrtc.go:111:25: Error return value of `c.preparePeerConnection` is not checked (errcheck)
        c.preparePeerConnection(config)
                               ^
pkg/snowflake/lib/lib_test.go:54:24: field `toRelease` is unused (unused)
type FakePeers struct{ toRelease *WebRTCPeer }
                       ^
pkg/snowflake/lib/lib_test.go:249:48: S1019: should use make([]byte, 100001) instead (gosimple)
                                &MockTransport{http.StatusOK, make([]byte, 100001, 100001)},
                                                                           ^
pkg/snowflake/lib/lib_test.go:76:7: ineffectual assignment to err (ineffassign)
                        p, err = NewPeers(d)
                           ^
pkg/config/logger.go:28:13: Error return value of `os.MkdirAll` is not checked (errcheck)
        os.MkdirAll(Path, 0750)
                   ^
pkg/snowflake/bootstrap.go:79:9: Error return value of `f.Write` is not checked (errcheck)
        f.Write([]byte(torrc))
               ^
pkg/snowflake/bootstrap.go:23:7: const `torrcOrig` is unused (unused)
const torrcOrig = `UseBridges 1
      ^
pkg/vpn/menshen/gateway.go:104:13: Error return value of `errors.New` is not checked (errcheck)
                errors.New("obfs4 is not supported for v5 right now")
                          ^
```

</details>

With this trick the following command works:
```bash
go build -o foo ./gui
```

**Reply by ldez** [reply]
The patch:

```diff
diff --git i/gui/handlers.cpp w/gui/handlers.cpp
index 48d6d11a..621833e1 100644
--- i/gui/handlers.cpp
+++ w/gui/handlers.cpp
@@ -1,3 +1,4 @@
+//go:build cpp
 #include <QTimer>
 #include <QDebug>
 #include <QDesktopServices>
diff --git i/gui/main.cpp w/gui/main.cpp
index 207fc730..b2dbffe7 100644
--- i/gui/main.cpp
+++ w/gui/main.cpp
@@ -1,3 +1,4 @@
+//go:build cpp
 #include <csignal>
 #include <unistd.h>
 #include <QtGui/qfontdatabase.h>
diff --git i/gui/qjsonmodel.cpp w/gui/qjsonmodel.cpp
index 030a2d48..1addca7f 100644
--- i/gui/qjsonmodel.cpp
+++ w/gui/qjsonmodel.cpp
@@ -1,3 +1,4 @@
+//go:build cpp
 /*
  * The MIT License (MIT)
  *
```

You can use it inside your MR https://0xacab.org/leap/bitmask-vpn/-/merge_requests/271

**Reply by leap-pea** [reply]
Thank you so much, I will test it soon!

**Reply by leap-pea** [reply]
Using `//go:build cpp` works perfectly fine. Now I can just use `golangci-lint run`.
